### ISSUE TITLE
Full system freeze on missing semicolon

### ISSUE THREAD

**Issue by ghost** [rust-lang/rust#34375]
I've been experiencing a complete system freeze when I compile the following code:

```
    macro_rules! test_macro {
        () => {{
            use std::env;
            let e = env::var("_")
        }}
    }

    #[cfg(test)]
    mod tests {
        #[test]
        fn it_works() {
            test_macro!();
        }
    }
```

Note the missing semicolon at the end of the assignment statement. I have been reliably able to reproduce this issue on Ubuntu 14.04.x and 16.04. The issue doesn't seem to be caused by OOM conditions as the freeze is instantaneous and no heavy swapping occurs (swap is enabled), but I may be mistaken.

Rust version: 1.9
Arch: x86_64

**Comment by jseyfried** [comment]
cc me

**Comment by nagisa** [comment]
Both stable and nightly result in OOM here:

```
> ~/.cargo/bin/rustup run stable rustc test.rs --test
test.rs:5:5: 5:6 error: expected one of `.`, `;`, or an operator, found `}`
test.rs:5     }}
              ^
fatal runtime error: out of memory
> ~/.cargo/bin/rustup run nightly rustc test.rs --test
test.rs:5:5: 5:6 error: expected one of `.`, `;`, `?`, or an operator, found `}`
test.rs:5     }}
              ^
fatal runtime error: out of memory
```

**Comment by jseyfried** [comment]
c.f. #34543

**Comment by goffrie** [comment]
I believe this was fixed by #37246.
